src/pushprocessor.ts

```diff
@@ -433,8 +433,74 @@ export class PushProcessor {
         return PushProcessor.cachedGlobToRegex[glob];
     }
 
+    /**
+     * Parse a dotted key into parts following the escape rules.
+     *
+     * @param str - The key of the push rule condition: a dotted field.
+     * @returns The unescaped parts to fetch.
+     */
+    private partsFotDottedKey(str: string): string[] {
+        const result = [];
+
+        // The current field and whether the previous character was the escape
+        // character (a backslash).
+        let part = "";
+        let escaped = false;
+
+        // Iterate over each character, and decide whether to append to the current
+        // part (following the escape rules) or to start a new part (based on the
+        // field separator).
```

Comment:
Do we know how hot the paths are that run this code? We could be introducing a performance problem here, and might need to e.g. try for a regex-based solution.

Reply:
This would run once per push rule per message, so fairly hot. I briefly attempted a regular expression solution and couldn't get it working.

Which aspect are you concerned about? The string concatenation or the iterating each character or something else?

Reply:
It might well be fine, but the iteration per-character made me wonder what the perf impact would be. There's no guarantee that a regular expression would be better, but I imagine regex engines are going to be good at iterating characters so maybe.

If we had nice unit tests for this function it would make it nice and easy to swap out the impl :-)

Reply:
For a regex solution I was thinking something dumb like

```
const r = /\\\./g;
const key = "foo.bar\\.baz";
const tmpkey = key.replaceAll(r, "__PLACEHOLDER__");
const parts = tmpkey.split(".").map(p => p.replaceAll("__PLACEHOLDER__", "."));
```

I haven't thought about edge cases, double-escaping etc.

Reply:
No idea whether it would be faster, and I generally prefer real parsing to solve a real parsing problem, but I am just worried about introducing a perf bomb :-)

Reply:
I've usually found the double escaping to break horribly with regular expression solutions, will take another look though!

Reply:
Yeah, it does feel dodgy.

Reply:
I went through a few iterations.

| Commit | Rate | Comment  |
|---|---|---|
| develop | 1.0 | Baseline |
| 7e11af82506d0393233047276c24784496b0ee60 | 0.25 | Looping solution |
| 97fe8e944506be01c95437c76617186fff3d79e8 | 0.20 | Regex solution |
| 322913ec114aa91533dbaffe5c614b857b124e15 | 0.50 | Split on regex solution |

Let me know what you think of them. I suspect the latest one is the easiest to understand?

Reply:
(Those timings are fairly unscientific. 🤷 )

Reply:
Although it seems SonarCloud doesn't like the splitting solution...

Reply:
Given a good set of unit tests, we are reasonably confident that all 3 are correct, I'd go for the looping solution for 2 main reasons:

1. It's the most direct implementation of the problem, so we can feel most confident it does what is intended. The regexes feel like hacks in comparison.
2. It performs approximately as fast as the others.

But, I think you're closest to it, so would like to hear what you think.

In terms of whether the performance is good enough, I guess the point is that we need to do this functionality, so we're going to need to find a way to make it work, so choosing something that performs about as well as we can is our best first guess.

Reply:
(I reviewed the looping solution and am happy with it.)

Reply:
I had an ah-ha moment last night about this -- we should be cached this on the push rule and then we only need to evaluate it once per (`event_match`) rule, not once per event per rule.

With that in mind, I definitely think the direct implementation is easiest to understand (and I'm most confident it isn't hitting other edge-cases).

Reply:
Brilliant, good thinking.